src/stratagus/luacallback.cpp

```diff
@@ -49,8 +49,9 @@ LuaCallbackImpl::LuaCallbackImpl(lua_State *l, lua_Object f) : luastate(l)
 	}
 	lua_pushvalue(l, f);
 	luaref = luaL_ref(l, LUA_REGISTRYINDEX);
-	refcounter =
-		std::shared_ptr<void>(nullptr, [this](void*) { luaL_unref(luastate, LUA_REGISTRYINDEX, luaref); });
+	refcounter = std::shared_ptr<void>(nullptr, [lua = luastate, ref = luaref](void *) {
```

Comment:
I was just some time away from creating a PR to fix this lambda :-)

Reply:
Sad I don't catch it in my UTs :-/

Reply:
Here it was also not visible. Not even with valgrind.
Only gcc with asan did show it.

Reply:
The report looks like this:
```
./stratagus_tests --tc=Lua"*" -s
=================================================================
==17350==ERROR: AddressSanitizer: stack-use-after-scope on address 0x7f272ae0c298 at pc 0x561d6059c130 bp 0x7ffd84f2aea0 sp 0x7ffd84f2ae90
READ of size 4 at 0x7f272ae0c298 thread T0
    #0 0x561d6059c12f in operator() stratagus/src/stratagus/luacallback.cpp:53
    #1 0x561d6059e81a in _M_dispose /usr/lib/gcc/x86_64-pc-linux-gnu/13/include/g++-v13/bits/shared_ptr_base.h:527
    #2 0x561d6047f201 in std::_Sp_counted_base<(__gnu_cxx::_Lock_policy)2>::_M_release() /usr/lib/gcc/x86_64-pc-linux-gnu/13/include/g++-v13/bits/shared_ptr_base.h:346
    #3 0x561d605030bb in std::__shared_count<(__gnu_cxx::_Lock_policy)2>::~__shared_count() /usr/lib/gcc/x86_64-pc-linux-gnu/13/include/g++-v13/bits/shared_ptr_base.h:1071
    #4 0x561d6059ec95 in std::__shared_ptr<void, (__gnu_cxx::_Lock_policy)2>::~__shared_ptr() /usr/lib/gcc/x86_64-pc-linux-gnu/13/include/g++-v13/bits/shared_ptr_base.h:1524
    #5 0x561d6059ecb1 in std::shared_ptr<void>::~shared_ptr() /usr/lib/gcc/x86_64-pc-linux-gnu/13/include/g++-v13/bits/shared_ptr.h:175
    #6 0x561d6059da12 in LuaCallbackImpl::~LuaCallbackImpl() stratagus/src/stratagus/luacallback.cpp:193
    #7 0x561d6047a23b in DOCTEST_ANON_FUNC_8 stratagus/tests/stratagus/test_luacallback.cpp:136
    #8 0x561d604146b9 in doctest::Context::run() stratagus/third-party/doctest/doctest/doctest.h:7007
    #9 0x561d604163d7 in main stratagus/third-party/doctest/doctest/doctest.h:7085
    #10 0x7f272dd88ee9 in __libc_start_call_main ../sysdeps/nptl/libc_start_call_main.h:58
    #11 0x7f272dd88fa4 in __libc_start_main_impl ../csu/libc-start.c:360
    #12 0x561d603e3960 in _start (build/stratagus-asan/stratagus_tests+0x1ed960)

Address 0x7f272ae0c298 is located in stack of thread T0 at offset 664 in frame
    #0 0x561d60478908 in DOCTEST_ANON_FUNC_8 stratagus/tests/stratagus/test_luacallback.cpp:119

  This frame has 45 object(s):
    [48, 49) 'raii' (line 120)
    [64, 65) '<unknown>'
    [80, 84) '<unknown>'
    [96, 100) '<unknown>'
    [112, 116) '<unknown>'
    [128, 132) '<unknown>'
    [144, 148) '<unknown>'
    [160, 164) '<unknown>'
    [176, 180) '<unknown>'
    [192, 196) '<unknown>'
    [208, 212) '<unknown>'
    [224, 228) '<unknown>'
    [240, 244) '<unknown>'
    [256, 260) '<unknown>'
    [272, 276) '<unknown>'
    [288, 292) '<unknown>'
    [304, 308) '<unknown>'
    [320, 324) '<unknown>'
    [336, 340) '<unknown>'
    [352, 356) '<unknown>'
    [368, 372) '<unknown>'
    [384, 388) '<unknown>'
    [400, 408) '<unknown>'
    [432, 440) '<unknown>'
    [464, 472) '<unknown>'
    [496, 504) '<unknown>'
    [528, 536) '<unknown>'
    [560, 600) 'f' (line 123)
    [640, 680) 'g' (line 127) <== Memory access at offset 664 is inside this variable
    [720, 744) '<unknown>'
    [784, 808) '<unknown>'
    [848, 872) '<unknown>'
    [912, 936) '<unknown>'
    [976, 1000) '<unknown>'
    [1040, 1072) '<unknown>'
    [1104, 1136) '<unknown>'
    [1168, 1200) '<unknown>'
    [1232, 1264) '<unknown>'
    [1296, 1328) '<unknown>'
    [1360, 1392) '<unknown>'
    [1424, 1568) 'DOCTEST_RB' (line 128)
    [1632, 1776) 'DOCTEST_RB' (line 129)
    [1840, 1984) 'DOCTEST_RB' (line 131)
    [2048, 2192) 'DOCTEST_RB' (line 132)
    [2256, 2400) 'DOCTEST_RB' (line 135)
HINT: this may be a false positive if your program uses some custom stack unwind mechanism, swapcontext or vfork
      (longjmp and C++ exceptions *are* supported)
SUMMARY: AddressSanitizer: stack-use-after-scope stratagus/src/stratagus/luacallback.cpp:53 in operator()
Shadow bytes around the buggy address:
  0x7f272ae0c000: f1 f1 f1 f1 f1 f1 01 f2 f8 f2 f8 f2 f8 f2 f8 f2
  0x7f272ae0c080: f8 f2 f8 f2 f8 f2 f8 f2 f8 f2 f8 f2 f8 f2 f8 f2
  0x7f272ae0c100: f8 f2 f8 f2 f8 f2 f8 f2 f8 f2 f8 f2 f8 f2 f8 f2
  0x7f272ae0c180: f8 f2 f8 f2 f2 f2 f8 f2 f2 f2 f8 f2 f2 f2 f8 f2
  0x7f272ae0c200: f2 f2 f8 f2 f2 f2 00 00 00 00 00 f2 f2 f2 f2 f2
=>0x7f272ae0c280: f8 f8 f8[f8]f8 f2 f2 f2 f2 f2 f8 f8 f8 f2 f2 f2
  0x7f272ae0c300: f2 f2 f8 f8 f8 f2 f2 f2 f2 f2 f8 f8 f8 f2 f2 f2
  0x7f272ae0c380: f2 f2 f8 f8 f8 f2 f2 f2 f2 f2 f8 f8 f8 f2 f2 f2
  0x7f272ae0c400: f2 f2 f8 f8 f8 f8 f2 f2 f2 f2 f8 f8 f8 f8 f2 f2
  0x7f272ae0c480: f2 f2 f8 f8 f8 f8 f2 f2 f2 f2 f8 f8 f8 f8 f2 f2
  0x7f272ae0c500: f2 f2 f8 f8 f8 f8 f2 f2 f2 f2 f8 f8 f8 f8 f2 f2
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==17350==ABORTING

```